File: src/config/paths.ts
import { join, dirname, resolve, isAbsolute } from "path";
import { fileURLToPath } from "url";
import { existsSync, readFileSync } from "fs";

const __dirname = dirname(fileURLToPath(import.meta.url));

const projectRoot = resolve(__dirname, "..", "..");
const homeMarker = join(projectRoot, ".podcli-home");
const dataDir = resolve(process.env.PODCLI_DATA || join(projectRoot, "data"));
const outputDir = resolve(process.env.PODCLI_OUTPUT || join(dataDir, "output"));

function resolveHome(): string {
  if (process.env.PODCLI_HOME) {
    return resolve(process.env.PODCLI_HOME);
  }
  if (existsSync(homeMarker)) {
    try {
      const marker = readFileSync(homeMarker, "utf-8").trim();
      if (marker) {
        return isAbsolute(marker) ? resolve(marker) : resolve(projectRoot, marker);
      }
    } catch {
      // Unreadable marker — fall through to the default home.
    }
  }
  return resolve(projectRoot, ".podcli");
}

const home = resolveHome();

// Hermetic installs run from a bundled runtime where the backend lives outside
// projectRoot; the launcher points PODCLI_BACKEND at it.
function resolveBackendDir(): string {
  if (process.env.PODCLI_BACKEND) {
    return resolve(process.env.PODCLI_BACKEND);
  }

  const sourceBackend = join(projectRoot, "backend");
  if (existsSync(join(sourceBackend, "cli.py"))) {
    return sourceBackend;
  }

  const runtimeBackend = join(projectRoot, "runtime", "backend");
  if (existsSync(join(runtimeBackend, "cli.py"))) {
    return runtimeBackend;
  }

  return sourceBackend;
}

const backendDir = resolveBackendDir();

function detectPython(): string {
  if (process.env.PYTHON_PATH) return process.env.PYTHON_PATH;
  const isWindows = process.platform === "win32";
  const venvPython = join(
    projectRoot,
    "venv",
    isWindows ? "Scripts" : "bin",
    isWindows ? "python.exe" : "python3"
  );
  if (existsSync(venvPython)) return venvPython;
  return isWindows ? "python" : "python3";
}

export const paths = {
  home,
  projectRoot,
  backendDir,
  homeMarker,
  dataDir,
  cache: join(dataDir, "cache"),
  transcripts: join(dataDir, "cache", "transcripts"),
  packed: join(home, "packed"),
  working: join(dataDir, "working"),
  output: outputDir,
  logs: join(dataDir, "logs"),
  assets: join(home, "assets"),
  assetsRegistry: join(home, "assets", "registry.json"),
  history: join(home, "history"),
  clipsHistory: join(home, "history", "clips.json"),
  knowledge: join(home, "knowledge"),
  uiState: join(home, "ui-state.json"),
  corrections: join(home, "corrections.json"),
  thumbnailConfig: join(home, "thumbnail-config.json"),
  integrations: join(home, "integrations.json"),
  pythonBackend: join(backendDir, "main.py"),
  pythonPath: detectPython(),
  ffmpegPath: process.env.FFMPEG_PATH || "ffmpeg",
  ffprobePath: process.env.FFPROBE_PATH || "ffprobe",
};

// Python falls back to the host locale encoding, which is cp1252 on most Windows
// installs. Titles carry em dashes and smart quotes, and U+201D encodes byte 0x9D,
// which cp1252 leaves undefined: reading it raises UnicodeDecodeError. Every spawn
// of the Python backend goes through here so no call site can forget.
export function pythonEnv(extra: Record<string, string> = {}): NodeJS.ProcessEnv {
  return {
    ...process.env,
    PYTHONUNBUFFERED: "1",
    PYTHONIOENCODING: "utf-8",
    PYTHONUTF8: "1",
    ...extra,
  };
}

File: backend/config/paths.py
from __future__ import annotations

import os
from pathlib import Path


def _project_root() -> Path:
    return Path(__file__).resolve().parents[2]


def _marker_path() -> Path:
    return _project_root() / ".podcli-home"


def _read_home_marker() -> str | None:
    marker = _marker_path()
    if not marker.exists():
        return None
    try:
        value = marker.read_text(encoding="utf-8").strip()
    except OSError:
        return None
    return value or None


def _resolve_home() -> Path:
    env_home = os.environ.get("PODCLI_HOME")
    if env_home:
        return Path(env_home).expanduser().resolve()

    marker_home = _read_home_marker()
    if marker_home:
        marker_path = Path(marker_home).expanduser()
        if not marker_path.is_absolute():
            marker_path = (_project_root() / marker_path).resolve()
        else:
            marker_path = marker_path.resolve()
        return marker_path

    return (_project_root() / ".podcli").resolve()


def _build_paths() -> dict[str, str]:
    home = _resolve_home()
    project_root = _project_root()
    data_dir = Path(os.environ.get("PODCLI_DATA", str(project_root / "data"))).expanduser().resolve()
    return {
        "home": str(home),
        "project_root": str(project_root),
        "cache": str(data_dir / "cache"),
        "transcripts": str(data_dir / "cache" / "transcripts"),
        "packed": str(home / "packed"),
        "working": str(data_dir / "working"),
        "output": str(data_dir / "output"),
        "logs": str(data_dir / "logs"),
        "assets": str(home / "assets"),
        "assetsRegistry": str(home / "assets" / "registry.json"),
        "history": str(home / "history"),
        "clipsHistory": str(home / "history" / "clips.json"),
        "knowledge": str(home / "knowledge"),
        "uiState": str(home / "ui-state.json"),
        "thumbnailConfig": str(home / "thumbnail-config.json"),
        "corrections": str(home / "corrections.json"),
        "integrations": str(home / "integrations.json"),
        "profileMarker": str(_marker_path()),
    }


paths = _build_paths()


def reload_paths() -> dict[str, str]:
    """Recompute paths in place after the active-home marker changes, so existing
    `from config.paths import paths` references reflect the new home without a
    process restart."""
    paths.clear()
    paths.update(_build_paths())
    return paths

